src/realm/util/logger.cpp

```diff
@@ -20,43 +20,98 @@
 
 #include <iostream>
 #include <mutex>
+#include <map>
 
 namespace realm::util {
 
 namespace {
 auto& s_logger_mutex = *new std::mutex;
 std::shared_ptr<util::Logger> s_default_logger;
-std::atomic<Logger::Level> s_default_level = Logger::Level::info;
 } // anonymous namespace
 
-void Logger::set_default_logger(std::shared_ptr<util::Logger> logger) noexcept
+size_t LogCategory::s_next_index = 0;
+static std::map<std::string_view, LogCategory*> log_catagory_map;
+
+LogCategory LogCategory::realm("Realm", nullptr);
+LogCategory LogCategory::storage("Storage", &realm);
+LogCategory LogCategory::transaction("Transaction", &storage);
+LogCategory LogCategory::query("Query", &storage);
+LogCategory LogCategory::object("Object", &storage);
+LogCategory LogCategory::notification("Notification", &storage);
+LogCategory LogCategory::sync("Sync", &realm);
+LogCategory LogCategory::client("Client", &sync);
+LogCategory LogCategory::session("Session", &client);
+LogCategory LogCategory::changeset("Changeset", &client);
+LogCategory LogCategory::network("Network", &client);
+LogCategory LogCategory::reset("Reset", &client);
+LogCategory LogCategory::server("Server", &sync);
+LogCategory LogCategory::app("App", &realm);
+LogCategory LogCategory::sdk("SDK", &realm);
+
+
+LogCategory::LogCategory(std::string_view name, LogCategory* parent)
+    : m_index(s_next_index++)
+    , m_default_level(Logger::Level::info)
 {
-    std::lock_guard lock(s_logger_mutex);
-    s_default_logger = logger;
+    if (parent) {
+        m_name = parent->get_name() + ".";
+        parent->m_children.push_back(this);
+    }
+    m_name += name;
+    log_catagory_map.emplace(m_name, this);
 }
 
-std::shared_ptr<util::Logger>& Logger::get_default_logger() noexcept
+LogCategory& LogCategory::get_category(std::string_view name)
+{
+    return *log_catagory_map.at(name); // Throws
+}
+
+void LogCategory::set_default_level_threshold(Level level)
 {
+    m_default_level.store(level);
+    for (auto c : m_children) {
+        c->set_default_level_threshold(level);
```

Comment:
would it be expected that the order of calls matters? This will override every child category up and down, no?

Reply:
Yes, that was the intention.